### PR TITLE
Fix/rotating drift correction

### PR DESCRIPTION
**Pull request by harmishhk**
Adds following optional features:
Rotating at start towards navigation plan (with drift correction)
Interpolate global plan
Reset orientations in global plan

### CODE REVIEW

File: src/rotating_controller.cpp [hunk 1]
```diff
@@ -3,74 +3,132 @@
...
-}
+    tf::StampedTransform initial_to_base, goal_to_base, base_in_maop;
+    tf_->lookupTransform(robot_base_frame_, initial_pose_.header.frame_id, ros::Time(0), initial_to_base);
+    tf_->lookupTransform(robot_base_frame_, goal_pose_.header.frame_id, ros::Time(0), goal_to_base);
+
+    initial_pose_tf = initial_to_base * initial_pose_tf;
+    goal_pose_tf = goal_to_base * goal_pose_tf;
+
+    // Differences to compansate while rotating
+    double x_diff = initial_pose_tf.getOrigin().x();
```
Comment (hogabrie) [inline src/rotating_controller.cpp]: We should follow some kind of naming convention, that makes it easier to read. This here seems to be in odom.
e.g. initial_pose_in_odom, O_pose_init

File: src/rotating_controller.cpp [hunk 2]
```diff
@@ -3,74 +3,132 @@
...
+    return false;
+  } catch (tf::ExtrapolationException& ex) {
+    ROS_ERROR("TF Extrapolation Error: %s\n", ex.what());
+    return false;
+  }
+}
+
+bool RotatingController::isGoalPoseReached(const tf::Stamped<tf::Pose>& robot_pose) {
+  try {
+    tf::StampedTransform goal_to_base;
```
Comment (hogabrie) [inline src/rotating_controller.cpp]: Also here it is very hard to follow due do multiple calculations on the same variable.

File: include/teb_local_planner/rotating_controller.h [hunk 3]
```diff
@@ -8,34 +8,36 @@
 namespace teb_local_planner {
 
 class RotatingController {
-public:
+ public:
   RotatingController() = default;
   virtual ~RotatingController() = default;
 
-  void initialize(ros::NodeHandle &nh);
+  void initialize(ros::NodeHandle& nh, tf::TransformListener*& tf, std::string& robot_base_frame);
```
[the remark this answers is not in the document]
Comment (harmishhk) [reply include/teb_local_planner/rotating_controller.h]: thanks, yes, in initialize method it does not make much sense, as class object is not reference. I guess, I was first trying to pass it at constructor

File: src/rotating_controller.cpp [hunk 4]
```diff
@@ -3,74 +3,132 @@
...
-      2.0, "Moving to desired goal orientation %.2f with anuglar velocity %.2f",
-      goal_th_, v_theta_samp);
-  cmd_vel.linear.x = 0;
-  cmd_vel.linear.y = 0;
-  cmd_vel.angular.z = v_theta_samp;
-  return true;
-}
+    tf::StampedTransform initial_to_base, goal_to_base, base_in_maop;
+    tf_->lookupTransform(robot_base_frame_, initial_pose_.header.frame_id, ros::Time(0), initial_to_base);
+    tf_->lookupTransform(robot_base_frame_, goal_pose_.header.frame_id, ros::Time(0), goal_to_base);
```
[the remark this answers is not in the document]
Comment (harmishhk) [reply src/rotating_controller.cpp]: AFAIU, `ros::Time(0)` is a placeholder to get latest available transform (not at `ros::Time::now()`), isn't that?
[http://wiki.ros.org/tf/Tutorials/tf and Time](http://wiki.ros.org/tf/Tutorials/tf%20and%20Time%20%28C%2B%2B%29)

[the remark this answers is not in the document]
Comment (remod) [reply src/rotating_controller.cpp]: Exactly, the latest available might be at different time stamps as initial_pose_.header.frame_id might be different from goal_pose_.header.frame_id

File: src/rotating_controller.cpp (hunk 1 above)
Comment (harmishhk) [reply src/rotating_controller.cpp]: Yes, I like the first option, I also used `_tf` for differentiating same between `geometry_msgs::Pose` and `tf::Pose`. I will change names in the code and update the pr.

**Comment by harmishhk** [comment]
Hi, I have updated the pull request to address all the comments. Could you please check it again?

File: src/rotating_controller.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (remod) [reply src/rotating_controller.cpp]: Any update on that?

File: src/teb_local_planner_ros.cpp [hunk 5]
```diff
@@ -283,31 +284,26 @@ bool TebLocalPlannerROS::setPlan(const std::vector<geometry_msgs::PoseStamped>&
...
   // setup turing on the spot at start, when asked
   if (cfg_.trajectory.turn_before_navigate && !global_plan_.empty()) {
-    auto &new_goal_ = global_plan_.back().pose;
-    bool same_goal = new_goal_.position.x == last_goal_.position.x &&
-                     new_goal_.position.y == last_goal_.position.y &&
-                     new_goal_.position.z == last_goal_.position.z &&
-                     new_goal_.orientation.x == last_goal_.orientation.x &&
-                     new_goal_.orientation.y == last_goal_.orientation.y &&
-                     new_goal_.orientation.z == last_goal_.orientation.z &&
+    auto& new_goal_ = global_plan_.back().pose;
```
Comment (remod) [inline src/teb_local_planner_ros.cpp]: Use const& for variables you are not intending to change?

File: src/rotating_controller.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (harmishhk) [reply src/rotating_controller.cpp]: I just need latest transforms from the tf, it is not necessary to have them strictly sychronized, so this is fine in my view.

File: src/teb_local_planner_ros.cpp (hunk 5 above)
Comment (harmishhk) [reply src/teb_local_planner_ros.cpp]: Updated PR with `const` variables

File: src/rotating_controller.cpp (hunk 1 above)
Comment (hogabrie) [reply src/rotating_controller.cpp]: I still think it would be easier to follow if the names were more descriptive.
goal_pose_tf = goal_to_base * goal_pose_tf; 
If you work on the same variables but then rotate them into different frames its almost impossible to track the frames.

Comment (harmishhk) [reply src/rotating_controller.cpp]: I have updated the PR with descriptive names, but here it does not show up. It only shows that file is `outdated`. Could you please check on the latest commit to see if that is sufficient. For exampe eaxtly the lines [58,59](https://github.com/ANYbotics/teb_local_planner/blob/00ac96c5647f06979fa9e5a831eccb02d07db2e9/src/rotating_controller.cpp#L58-L59) which you have added comments here.

Comment (hogabrie) [reply src/rotating_controller.cpp]: Ah sorry. Yes these poses look good now :)
But there is still robot_vel or vel_x which is kind of unclear. 
In loco for example we are really clear about this. E.g.
`positionWorldToBaseInBaseFrame = orientationWorldToBase.rotate(positionWorldToBaseInWorldFrame);
`

Comment (hogabrie) [reply src/rotating_controller.cpp]: I am a bit picky on that, I know. But I have seen so many errors in coordinate transformations in the past, because of unclear variable names. Even very popular packages like hector_gazebo_plugin have wrong transformations that we had to fix on our fork.

Comment (harmishhk) [reply src/rotating_controller.cpp]: The planner always works in given `robot_base_frame` which can be changed to anything, e.g. `feetcenter`. So, for me it is clear once everything is in specified robot base, calculations are just absolute. But if it makes clear I could add `_base` to all variables.

Comment (hogabrie) [reply src/rotating_controller.cpp]: Or add it above the block as a comment, I agree that adding _base to everything is an overkill.

Comment (harmishhk) [reply src/rotating_controller.cpp]: Added one more committ to the PR clarifying velocity frame in comments.
